### PR TITLE
StorageProxy implementation

### PR DESCRIPTION
**Pull request by bayesianmind**
- ThreadSafe operations.
- Fully suspending, coroutine compatible.
- Handle operations moved to suspend functions as well since
getParticleView always waits for syncronized state first.
- Moved handle to same layer as proxy to avoid circular imports.

### CODE REVIEW

File: java/arcs/core/storage/StorageProxy.kt [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+                store.setCallback(ProxyCallback<Data, Op, T>(::onMessage))
+                listenerAttached = true
+            }
+
+            handles.add(handle)
+
+            // Change to synchronized mode as soon as we get any read handles and send a request to get
+            // the full model (once).
+            if (!this.hasReaders) {
+                this.requestSynchronization()
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageProxy.kt]: Another deadlocking opportunity - the store may call back into `onMessage` before this `withLock` has released.

Comment (jasonwyatt) [reply java/arcs/core/storage/StorageProxy.kt]: I think it's probably fine to use a different mutex or an atomic boolean here as well.

File: java/arcs/core/storage/StorageProxy.kt [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+     */
+    suspend fun registerHandle(handle: Handle<Data, Op, T>): VersionMap {
+        // non-readers don't get callbacks, return early
+        if (!handle.canRead) {
+            mutex.withLock {
+                return crdt.versionMap.copy()
+            }
+        }
+
+        mutex.withLock() {
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageProxy.kt]: It seems like the only reason we need to hold a lock here is to update the `handles` set and get the current version.  I think most of the chunks within can be brought outside the lock to remove the risk of deadlock.  Additionally, something like `requestSynchronization` might be better done as something which is self-contained and returns a deferred with whatever you need from it - rather than having it be a method with an opaque side-effect.

File: java/arcs/core/storage/StorageProxy.kt [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+            if (handle.options.notifySync && this.synchronized) {
+                coroutineScope {
+                    launch { handle.callback?.onSync() }
+                }
+            }
+            return crdt.versionMap.copy()
+        }
+    }
+
+    suspend fun deregisterHandle(handle: Handle<Data, Op, T>) {
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageProxy.kt]: Public functions should have kdoc.

File: java/arcs/core/storage/StorageProxy.kt [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+        mutex.withLock {
+            handles.remove(handle)
+        }
+    }
+    /**
+    * Apply a CRDT operation to the [CrdtModel] that this [StorageProxy] manages, notifies read
+     * handles, and forwards the write to the [Store].
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageProxy.kt]: Whoops - wonky indentation

File: java/arcs/core/storage/StorageProxy.kt [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+     */
+    suspend fun applyOp(op: Op): Boolean {
+        mutex.withLock {
+            if (!crdt.applyOperation(op)) {
+                return false
+            }
+            val msg = ProxyMessage.Operations<Data, Op, T>(listOf(op), null)
+            store.onProxyMessage(msg)
+            notifyUpdate(op)
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageProxy.kt]: this should be done outside of the lock, to avoid deadlocking issues

File: java/arcs/core/storage/StorageProxy.kt [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+     */
+    suspend fun getParticleView(): ValueAndVersion<T> {
+        val future = CompletableDeferred<ValueAndVersion<T>>()
+        mutex.withLock {
+            if (synchronized) {
+                return ValueAndVersion(crdt.consumerView, crdt.versionMap)
+            }
+            waitingSyncs.add(future)
+        }
+        return future.await()
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageProxy.kt]: Could be simpler and less brittle as:

```kotlin
val future = CompletableDeferred<ValueAndVersion<T>>()
mutex.withLock {
    if (synchronized) future.resolve(ValueAndVersion(crdt.consumerView, crdt.versionMap.copy()))
    else waitingSyncs.add(future)
}
return future.await()
```

Also, I ctrl+f'd and looked for where `waitingSyncs` is purged, and didn't see anything.

File: java/arcs/core/storage/StorageProxy.kt [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+    }
+
+    suspend fun onMessage(message: ProxyMessage<Data, Op, T>): Boolean {
+        mutex.withLock {
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageProxy.kt]: The fact that this mutex locks the entire body - which also goes off and does other things that could risk a re-entrant deadlock tells me this should probably be re-swizzled.

File: java/arcs/core/storage/StorageProxy.kt [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+                    val modelUpdate = ProxyMessage.ModelUpdate<Data, Op, T>(this.crdt.data, null)
+                    this.store.onProxyMessage(modelUpdate);
+                }
+                else -> {
+                    throw CrdtException("Invalid operation provided to onMessage, message: $message");
+                }
+            }
+            return true;
+        }
+    }
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageProxy.kt]: Lots of semicolons... what is this, Java?  :stuck_out_tongue:

File: java/arcs/core/storage/StorageProxy.kt [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+            this.notifySync();
+        }
+    }
+
+    private suspend fun clearSynchronized() {
+        if (this.synchronized) {
+            this.synchronized = false;
+            this.notifyDesync();
+        }
+    }
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageProxy.kt]: More semicolons...

File: java/arcs/core/storage/StorageProxy.kt [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+                launch {
+                    handle.callback?.onDesync()
+                }
+            }
+        }
+    }
+
+    private suspend fun requestSynchronization(): Boolean {
+        val msg = ProxyMessage.SyncRequest<Data, Op, T>(null)
+        return this.store.onProxyMessage(msg);
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageProxy.kt]: semicolon

File: javatests/arcs/core/storage/StorageProxyTest.kt [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+    @Mock private lateinit var mockCrdtOperation: CrdtOperation
+    @Mock private lateinit var mockCrdtModel: CrdtModel<CrdtData, CrdtOperation, String>
+    @Mock private lateinit var mockCrdtData: CrdtData
+
+    @Before
+    fun setup() {
+        MockitoAnnotations.initMocks(this)
+        Mockito.`when`(mockStorageEndpointProvider.getStorageEndpoint()).thenReturn(mockStorageEndpoint)
+        Mockito.`when`(mockCrdtModel.data).thenReturn(mockCrdtData)
+        Mockito.`when`(mockCrdtModel.versionMap).thenReturn(VersionMap())
```
Comment (jasonwyatt) [inline javatests/arcs/core/storage/StorageProxyTest.kt]: Use `whenever` from `mockito-kotlin`

File: javatests/arcs/core/storage/StorageProxyTest.kt [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+import org.mockito.Mockito
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.verify
+import org.mockito.MockitoAnnotations
+
+/** Tests for [StorageProxy]. */
+@Suppress("UNCHECKED_CAST", "UNUSED_VARIABLE")
+@ExperimentalCoroutinesApi
+@RunWith(JUnit4::class)
+class StorageProxyTest {
```
Comment (jasonwyatt) [inline javatests/arcs/core/storage/StorageProxyTest.kt]: I think we need some tests here which try to cause deadlocks / test 'randomized' concurrent operations

File: javatests/arcs/core/storage/handle/CollectionImplTest.kt [hunk 13]
```diff
@@ -28,61 +37,69 @@ class CollectionImplTest {
...
+            Set<MockDataItem>>
+    @Mock private lateinit var mockStorageEndpointProvider:
+        StorageCommunicationEndpointProvider<CrdtSet.Data<MockDataItem>,
+            CrdtSet.IOperation<MockDataItem>, Set<MockDataItem>>
+
     @Before
     fun setUp() {
         MockitoAnnotations.initMocks(this)
-        val storageProxy = StorageProxy(CrdtSet<MockDataItem>())
+        `when`(mockStorageEndpointProvider.getStorageEndpoint()).thenReturn(mockStorageEndpoint)
```
Comment (jasonwyatt) [inline javatests/arcs/core/storage/handle/CollectionImplTest.kt]: use `whenever`

File: java/arcs/core/storage/Handle.kt [hunk 14]
```diff
@@ -21,10 +21,20 @@ import arcs.core.crdt.internal.VersionMap
  *
  * This is mostly a place-holder at the moment.
```
[the remark this answers is not in the document]
Comment (bayesianmind) [reply java/arcs/core/storage/Handle.kt]: No I don't think so, I'll remove that.

File: java/arcs/core/storage/Handle.kt [hunk 15]
```diff
@@ -21,10 +21,20 @@ import arcs.core.crdt.internal.VersionMap
...
  */
-interface Callbacks {
-    fun onUpdate(handle: Handle<*, *, *>)
+interface Callbacks<Op : CrdtOperation> {
+    fun onUpdate(op: Op)
+    fun onSync()
+    fun onDesync()
 }
 
+data class HandleOptions(
```
[the remark this answers is not in the document]
Comment (bayesianmind) [reply java/arcs/core/storage/Handle.kt]: I'm thinking about removing these entirely. If someone doesn't want to be notified for syncs, they can just have an empty notifySync().

File: java/arcs/core/storage/Handle.kt [hunk 16]
```diff
@@ -21,10 +21,20 @@ import arcs.core.crdt.internal.VersionMap
  *
  * This is mostly a place-holder at the moment.
  */
-interface Callbacks {
-    fun onUpdate(handle: Handle<*, *, *>)
+interface Callbacks<Op : CrdtOperation> {
+    fun onUpdate(op: Op)
+    fun onSync()
+    fun onDesync()
```
[the remark this answers is not in the document]
Comment (bayesianmind) [reply java/arcs/core/storage/Handle.kt]: For sure. Done.

File: javatests/arcs/core/storage/StorageProxyTest.kt (hunk 12 above)
Comment (bayesianmind) [reply javatests/arcs/core/storage/StorageProxyTest.kt]: Agreed, I'll work on it

File: java/arcs/core/storage/StorageProxy.kt [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+import arcs.core.crdt.CrdtOperation
+import arcs.core.crdt.internal.VersionMap
+import arcs.core.util.guardWith
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.sync.Mutex
+import kotlinx.coroutines.sync.withLock
+
+/** ValueAndVersion is a tuple of a value of some type and a [VersionMap] provided from its [StorageProxy]. */
```
Comment (csilvestrini) [inline java/arcs/core/storage/StorageProxy.kt]: Square brackets around types: `[ValueAndVersion]`

But even better is to omit it entirely, since it's obvious what you're talking about: `A tuple of ...`

File: java/arcs/core/storage/StorageProxy.kt [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+data class ValueAndVersion<T>(val value: T, val versionMap: VersionMap)
+
+/**
+ * StorageProxy is an intermediary between a [Handle] and [Store].
```
Comment (csilvestrini) [inline java/arcs/core/storage/StorageProxy.kt]: Same comment here: `An intermediary between ...`

File: java/arcs/core/storage/StorageProxy.kt [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+ * It provides up-to-date CRDT state to readers, and ensures write operations apply cleanly before
+ * forwarding to the store.
+ *
+ * @param T the consumer data type for the model behind this proxy
+ * @property model a concrete instance of the [CrdtModel] the proxy will use to keep a local copy.
```
Comment (csilvestrini) [inline java/arcs/core/storage/StorageProxy.kt]: Should this be `@property initialCrdt` instead?

File: java/arcs/core/storage/StorageProxy.kt [hunk 20]
```diff
@@ -0,0 +1,221 @@
...
+                        }
+                        this.notifyUpdate(op);
+                    }
+                }
+                is ProxyMessage.SyncRequest -> {
+                    // storage wants our latest state
+                    val modelUpdate = ProxyMessage.ModelUpdate<Data, Op, T>(this.crdt.data, null)
+                    this.store.onProxyMessage(modelUpdate);
+                }
+                else -> {
```
Comment (csilvestrini) [inline java/arcs/core/storage/StorageProxy.kt]: No need for the curly braces here. This could just be:

```
else -> throw CrdtException(...)
```

If you need to line break:

```
else -> throw CrdtException(
    "....."
)
```

File: javatests/arcs/core/storage/BUILD [hunk 21]
```diff
@@ -7,8 +7,11 @@ licenses(["notice"])
 
 package(default_visibility = ["//visibility:public"])
 
+TEST_SRCS = glob(["*Test.kt"])
+
 arcs_kt_android_test_suite(
     name = "storage",
+    srcs = TEST_SRCS,
```
Comment (csilvestrini) [inline javatests/arcs/core/storage/BUILD]: This isn't strictly needed, it will glob all *.kt sources automatically

File: javatests/arcs/core/storage/StorageProxyTest.kt [hunk 22]
```diff
@@ -0,0 +1,133 @@
...
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.junit.runners.JUnit4
+import org.mockito.Mock
+import org.mockito.Mockito
+import org.mockito.Mockito.mock
+import org.mockito.Mockito.verify
+import org.mockito.MockitoAnnotations
+
+/** Tests for [StorageProxy]. */
```
Comment (csilvestrini) [inline javatests/arcs/core/storage/StorageProxyTest.kt]: This isn't strictly needed -- it's pretty clear what class this tests, and IDEs will let you auto-switch

File: javatests/arcs/core/storage/StorageProxyTest.kt [hunk 23]
```diff
@@ -0,0 +1,133 @@
...
+        val view = storageProxy.getParticleView()
+
+        assertThat(view.value).isEqualTo("someData")
+        assertThat(view.versionMap).isEqualTo(VersionMap())
+    }
+
+    private fun handle(name: String,
+                       storageProxy: StorageProxy<CrdtData, CrdtOperation, String>,
+                       reader: Boolean):
+        Handle<CrdtData, CrdtOperation, String> {
```
Comment (csilvestrini) [inline javatests/arcs/core/storage/StorageProxyTest.kt]: Indenting:

```
private fun handle(
    name: String,
    storageProxy: StorageProxy<CrdtData, CrdtOperation, String>,
    reader: Boolean
): Handle<CrdtData, CrdtOperation, String> {
```

File: javatests/arcs/core/storage/StorageProxyTest.kt [hunk 24]
```diff
@@ -0,0 +1,133 @@
...
+        storageProxy.registerHandle(readHandle)
+        assertThat(storageProxy.onMessage(ProxyMessage.Operations(listOf(mockCrdtOperation), null)))
+            .isTrue()
+        val view = storageProxy.getParticleView()
+
+        assertThat(view.value).isEqualTo("someData")
+        assertThat(view.versionMap).isEqualTo(VersionMap())
+    }
+
+    private fun handle(name: String,
```
Comment (csilvestrini) [inline javatests/arcs/core/storage/StorageProxyTest.kt]: `newHandle` or `createHandle`

File: javatests/arcs/core/storage/StorageProxyTest.kt [hunk 25]
```diff
@@ -0,0 +1,133 @@
...
+                       storageProxy: StorageProxy<CrdtData, CrdtOperation, String>,
+                       reader: Boolean):
+        Handle<CrdtData, CrdtOperation, String> {
+        val handle = Handle(name, storageProxy, reader)
+        handle.callback = mock(Callbacks::class.java) as Callbacks<CrdtOperation>
+        return handle
```
Comment (csilvestrini) [inline javatests/arcs/core/storage/StorageProxyTest.kt]: A cool Kotlin trick you might not know about:

```
return Handle(...).apply {
    callback = mock(...)
}
```

This lets you create and return it all in one expression. You could also omit the curly braces from the function (and return type) if you wanted:

```
fun newHandle(...) = Handle(...).apply { ... }
```

File: java/arcs/core/storage/StorageProxy.kt [hunk 26]
```diff
@@ -0,0 +1,221 @@
...
+            store.onProxyMessage(msg)
+            notifyUpdate(op)
+            return true
+        }
+    }
+
+    /**
+     * Return the current local version of the model, as well as the current associated version
+     * map for the data.
+     * Suspends until it has a synchronized view of the data.
```
Comment (csilvestrini) [inline java/arcs/core/storage/StorageProxy.kt]: Either insert a blank line to separate the paragraphs, or just combine into one paragraph

**Comment by csilvestrini** [comment]
Great job! I just had style nits

File: java/arcs/core/storage/handle/CollectionImpl.kt [hunk 27]
```diff
@@ -27,27 +30,22 @@ private typealias SetHandle<T> = Handle<CrdtSet.Data<T>, CrdtSet.IOperation<T>,
 class CollectionImpl<T : Referencable>(
```
Comment (galganif) [inline java/arcs/core/storage/handle/CollectionImpl.kt]: in ts this T is not Referenceable, rather it takes entity instances that are serialized to make sure they have an ID. do we have something like that here?

File: java/arcs/core/storage/StorageProxy.kt [hunk 28]
```diff
@@ -0,0 +1,227 @@
...
+                }
+            }
+            return true
+        }
+    }
+
+    // must be called under single-thread-confinement
+    private suspend fun setSynchronized() {
+        for (sync in waitingSyncs) {
+            sync.complete(getParticleView())
```
[the remark this answers is not in the document]
Comment (bayesianmind) [reply java/arcs/core/storage/StorageProxy.kt]: Whoops, good catch. I am adding a test for this case.

File: java/arcs/core/storage/handle/CollectionImpl.kt (hunk 27 above)
Comment (bayesianmind) [reply java/arcs/core/storage/handle/CollectionImpl.kt]: No idea-  I think jibbl might have the context on that but he's out sick. @jasonwyatt do you know?

File: java/arcs/core/storage/StorageProxy.kt (hunk 17 above)
Comment (bayesianmind) [reply java/arcs/core/storage/StorageProxy.kt]: Made the comment better:
/** [ValueAndVersion] is used to tie a value to the VersionMap it was current as-of. */

File: javatests/arcs/core/storage/BUILD (hunk 21 above)
Comment (bayesianmind) [reply javatests/arcs/core/storage/BUILD]: In a previous iteration I had a fakes class that I ended up removing. This mirrors the structure we use in the handle tests and is a better glob pattern since it only picks up test classes.

File: javatests/arcs/core/storage/StorageProxyTest.kt (hunk 22 above)
Comment (bayesianmind) [reply javatests/arcs/core/storage/StorageProxyTest.kt]: Awesome, glad we don't feel the need to exhaustively KDoc very obvious stuff.

File: javatests/arcs/core/storage/StorageProxyTest.kt (hunk 24 above)
Comment (bayesianmind) [reply javatests/arcs/core/storage/StorageProxyTest.kt]: Done. My 2.5 years of daily Go programming are leaking..

File: javatests/arcs/core/storage/StorageProxyTest.kt (hunk 25 above)
Comment (bayesianmind) [reply javatests/arcs/core/storage/StorageProxyTest.kt]: Neat!

File: java/arcs/core/storage/Handle.kt [hunk 29]
```diff
@@ -18,11 +18,25 @@ import arcs.core.crdt.internal.VersionMap
 /**
  * The [Callbacks] interface is a simple stand-in for the callbacks that a [Handle] might want
  * to use* to signal information back to the runtime.
```
[the remark this answers is not in the document]
Comment (bayesianmind) [reply java/arcs/core/storage/Handle.kt]: Yeah particle. Clarified: The [Callbacks] interface is a simple stand-in for the callbacks that a [Handle] might want  to use to signal information back to a subscriber (like a handle).

This is coming from Ray's thoughts that we've built too much coupling between these layers (I agree), and that the runtime should be the one wiring them up and deciding what's connected. This is also relevant when we talk about folks using Handles directly w/o particles.

File: javatests/arcs/core/storage/StorageProxyTest.kt (hunk 12 above)
Comment (bayesianmind) [reply javatests/arcs/core/storage/StorageProxyTest.kt]: I've upgraded AtomicFu so we can use their multi-platform ReentrantLock. That should mitigate the deadlock possibilities. I very much want some awesome randomized tests, but I'd like to do it as a follow-up CL integration test covering at minimum the collection handles + storage proxy + concurrent read/writes to said collection, the store fake (or even better a ram store).

File: java/arcs/core/storage/handle/CollectionImpl.kt (hunk 27 above)
Comment (jblebrun) [reply java/arcs/core/storage/handle/CollectionImpl.kt]: @jasonwyatt probably has better context. The use of `T : Referencable` here is done to match the type requirements on underlying CRDT type, which I believe was a decision made by him. 

Also possibly of note: these are the runtime-side handles; it may be the case that the SDK-side handles will have a slightly different shape.

Comment (jasonwyatt) [reply java/arcs/core/storage/handle/CollectionImpl.kt]: That's right, the Storage/CRDT stack is referencables-all-the-way-down - so to ensure things play nicely in the strongly typed world, all levels use Referencable as a lower bound on their type parameter.

File: java/arcs/core/storage/StorageProxy.kt [hunk 30]
```diff
@@ -0,0 +1,230 @@
...
+     */
+    suspend fun getParticleView(): ValueAndVersion<T> {
+        return getParticleViewAsync().await()
+    }
+
+    suspend fun getParticleViewAsync(): CompletableDeferred<ValueAndVersion<T>> {
+        val future = CompletableDeferred<ValueAndVersion<T>>()
+        mutex.withLock {
+            if (synchronized) {
+                return future.apply {
```
Comment (jasonwyatt) [inline java/arcs/core/storage/StorageProxy.kt]: Optional Nit: I typically prefer a single `return` at the bottom whenever possible, rather than short-circuiting early (unless the short-circuiting is due to a guard). It makes things slightly easier to read IMO.

File: javatests/arcs/core/storage/StorageProxyTest.kt (hunk 12 above)
Comment (jasonwyatt) [reply javatests/arcs/core/storage/StorageProxyTest.kt]: SGTM

File: javatests/arcs/core/storage/handle/CollectionImplTest.kt [hunk 31]
```diff
@@ -28,61 +37,75 @@ class CollectionImplTest {
...
     fun setUp() {
         MockitoAnnotations.initMocks(this)
-        val storageProxy = StorageProxy(CrdtSet<MockDataItem>())
+        whenever(mockStorageEndpointProvider.getStorageEndpoint()).thenReturn(mockStorageEndpoint)
+        // Obstensibly this should be a mock storageProxy, but since this unit test is primarily
+        // about testing CrdtOps we want at least a reasonable fake to apply our Crdt Ops.
+        val storageProxy = StorageProxy(mockStorageEndpointProvider, CrdtSet<MockDataItem>())
+        val makeProxySynced = ProxyMessage.ModelUpdate<CrdtSet.Data<MockDataItem>,
+            CrdtSet.IOperation<MockDataItem>, Set<MockDataItem>>(
+            model = CrdtSet.DataImpl(), id = null)
```
Comment (jasonwyatt) [inline javatests/arcs/core/storage/handle/CollectionImplTest.kt]: Style nit:

Parameters should go on separate lines when things need to wrap like this.  Also, given the huge generic type signature, I feel like it would be worthwhile to create some TypeAliases later (if necessary).

File: java/arcs/core/storage/Handle.kt [hunk 32]
```diff
@@ -44,36 +58,33 @@ interface Callbacks {
  * Implementations using this base are assumed to be based on a [CrdtModel] type. Concrete
  * subclasses should select a [CrdtData] and [CrdtOperation] type for the [Handle] when
  * implementing a concrete version of it, and expose only the consumer type [T].
+ *
+ * The base handle interface is at this layer to avoid circular dependencies.
  */
-abstract class Handle<Data : CrdtData, Op : CrdtOperation, T>(
+open class Handle<Data : CrdtData, Op : CrdtOperation, T>(
```
Comment (jblebrun) [inline java/arcs/core/storage/Handle.kt]: Why the change from `abstract` to `open`? Is there a case where it would be meaningful to instantiate the base class?

File: java/arcs/core/storage/Handle.kt [hunk 33]
```diff
@@ -44,36 +58,33 @@ interface Callbacks {
...
+     * [canRead] is whether this handle reads data so proxy can decide whether to keep its crdt
+     * up to date
+     */
+    val canRead: Boolean = true
 ) {
 
-    /** A list of current registered [Callbacks] instances */
-    private val callbacks = mutableListOf<Callbacks>()
+    /** The currently registered [Callbacks] instance */
+    var callback: Callbacks<Op>? = null
```
Comment (jblebrun) [inline java/arcs/core/storage/Handle.kt]: Does it make sense to set this on creation, and make it non-nullable?

File: java/arcs/core/storage/handle/CollectionImpl.kt (hunk 27 above)
Comment (galganif) [reply java/arcs/core/storage/handle/CollectionImpl.kt]: Ok, but there must be a place where the ID is added (if missing), right? Where is that done in this stack?

File: java/arcs/core/storage/Handle.kt (hunk 32 above)
Comment (bayesianmind) [reply java/arcs/core/storage/Handle.kt]: I use it in the storageproxy tests directly, since I don't want to depend on the subclasses impl.

Comment (jblebrun) [reply java/arcs/core/storage/Handle.kt]: It's pretty easy to create a one-off subclass for testing from this, no?
